File: 백준/Gold/1707. 이분 그래프/이분 그래프.cc
#include <iostream>
#include <vector>

using namespace std;

int k, v, e;
vector<vector<int>> graph(20005);
int visited[20005] = {0,};
bool answer = true;

// visited == 1: Red, visited == 2: Black
void dfs(int x, int rb) {
    for (int i = 0; i < graph[x].size(); i++) {
        if (visited[graph[x][i]] == 0) {
            int color = (rb % 2 == 1) ? 1 : 2;
            visited[graph[x][i]] = color;
            //cout << graph[x][i] << " ";
            dfs(graph[x][i], rb + 1);
        }
    }
}

int main(void) {
    cin.tie(NULL); ios_base::sync_with_stdio(false);
    
    cin >> k;
    
    for (int tc = 0; tc < k; tc++) {
        cin >> v >> e;
        answer = true;
        
        // 벡터 초기화
        for (int i = 1; i <= v; i++) {
            graph[i].clear();
        }
        fill(visited, visited + 20001, 0);
        
        for (int i = 0; i < e; i++) {
            int x, y;
            cin >> x >> y;
            graph[x].push_back(y);
            graph[y].push_back(x);
        }

        for (int i = 1; i <= v; i++) {
            if (visited[i] == 0) {
                visited[i] = 1;
                //cout << i << " ";
                dfs(i, 2); // dfs로 노드 방문하며 r-b-r-b-... 색칠
            }
            
        }
        
        for (int i = 1; i <= v; i++) {
            int color = visited[i];
            for (int j = 0; j < graph[i].size(); j++) {
                if (visited[graph[i][j]] == color) {
                    answer = false;
                    break;
                }
            }
            if (answer == false)
                break;
        }
        
        if (answer)
            cout << "YES\n";
        else
            cout << "NO\n";
    }

    return 0;
}

File: 백준/Gold/1707. 이분 그래프/이분 그래프.py
import sys
sys.setrecursionlimit(10**6)
k = int(sys.stdin.readline())
result = True

def dfs(x, before):
    for i in graph[x]:
        # 컬러 먼저 보기
        if color[i] == -1: # 컬러 부여 전
            color[i] = abs(before - 1)
        elif color[i] == before: # 같은 컬러 -> 같은 컬러 dfs : NO
           global result
           result = False

        # 방문 확인
        if visited[i] != 0:
            continue
        visited[i] = 1
        dfs(i, color[i])


for _ in range(k):
    v, e = map(int, sys.stdin.readline().split())
    graph = [[] for _ in range(v + 1)]
    color = [-1 for _ in range(v + 1)] # 0은 red, 1는 blue
    visited = [0 for _ in range(v + 1)]
    result = True

    for _ in range(e):
        a, b = map(int, sys.stdin.readline().split())
        graph[a].append(b)
        graph[b].append(a)

    for i in range(1, v + 1):
        if visited[i] != 0:
            continue
        visited[i] = 1
        dfs(i, 0)
    
    if result: print("YES")
    else: print("NO")